### DISCUSSION TITLE
`instanceof Promise` in loadable utility not reliable in all environments

### DISCUSSION THREAD

**Discussion by joseph-hanvy-sophos** [pmndrs/jotai#2498] [Bug report]
Recently faced an issue using jotai in a project that used some kind of es6.promise shim/polyfill to provide backwards compatibility for old browser.

It turns out the `instanceof Promise` here:
```
const isPromise = <Value>(
  x: unknown,
): x is Promise<Awaited<Value>> & PromiseMeta<Value> => x instanceof Promise
```
https://github.com/pmndrs/jotai/blob/v2.8.0/src/vanilla/utils/loadable.ts#L13

It seems it isn't a super reliable way to check - I replaced it with `typeof x.then === "function"` instead, but I'm not sure that's the best way to check for a promise either.

There's an interesting discussion here:
https://stackoverflow.com/a/28133130/1305699

I notice that `isPromiseLike` is used here, which works fine:
https://github.com/pmndrs/jotai/blob/9ceaef6403a9d932590b37852734bd8bd4958b74/src/react/useAtomValue.ts#L10

It might be best to use that everywhere instead of `instanceof Promise`, e.g.

https://github.com/pmndrs/jotai/blob/9ceaef6403a9d932590b37852734bd8bd4958b74/src/vanilla/utils/atomWithObservable.ts#L166
https://github.com/pmndrs/jotai/blob/9ceaef6403a9d932590b37852734bd8bd4958b74/src/vanilla/utils/unwrap.ts#L18
https://github.com/pmndrs/jotai/blob/9ceaef6403a9d932590b37852734bd8bd4958b74/src/vanilla/store.ts#L94
https://github.com/pmndrs/jotai/blob/9ceaef6403a9d932590b37852734bd8bd4958b74/src/vanilla/utils/atomWithObservable.ts#L166
https://github.com/pmndrs/jotai/blob/9ceaef6403a9d932590b37852734bd8bd4958b74/src/vanilla/utils/atomWithStorage.ts#L236

I doubt this is a problem faced by many but it should be an easy fix and it was a real pain to work out what was wrong!

**Comment by dai-shi** [comment]
> Recently faced an issue using jotai in a project that used some kind of es6.promise shim/polyfill to provide backwards compatibility for old browser.

So, in your environment, `new Promise() instanceof Promise` is `false`? That's unexpected. I use `PromiseLike` for api facing cases, but not everywhere. Did you experience the unexpected behavior with `loadable`, right? Maybe, `loadable` and `unwrap` should use PromiseLike.

**Reply by joseph-hanvy-sophos** [reply]
Interestingly `new Promise() isntanceof Promise` returns `true`, so it seems explicit promises 'pass', however in most use-cases it's the result of an async function so I checked `(async () => "hi")() instanceof Promise`, and that returns `false` in the environment (expect `true` and see that in just a new chrome window console)

We experienced the loadable atom never entering the loadable state, though it was version 2.5 so it was falling into this `if` check and immediately going to `hasData`:
```
        if (!(value instanceof Promise)) {
          return { state: 'hasData', data: value } as Loadable<Value>
        }
```
I'm upgrading to see if the same happens in later versions though, the code has changed but I assume it'll fail in the same way here:
https://github.com/pmndrs/jotai/blob/v2.8.0/src/vanilla/utils/loadable.ts#L45

**Reply by dai-shi** [reply]
Do you know any way to experiment the behavior in node or latest chrome?

`(async () => "hi")() instanceof Promise` being `false` is understandable.

**Reply by joseph-hanvy-sophos** [reply]
Latest chrome just using the console window works as I'd expect, with it returning `true`:

<img width="449" alt="image" src="https://github.com/pmndrs/jotai/assets/133230942/aa45fc97-dc57-4b41-bfd1-0de3ec518b75">

**Reply by joseph-hanvy-sophos** [reply]
Hmm, actually I'm not sure this is an issue, I'm having trouble recreating it stepping into that if condition

**Reply by joseph-hanvy-sophos** [reply]
Ah, it was an issue we found in our custom version of `loadable` that we created for setters that had a very similar structure, but used the result of `set` instead of `get`.

I'd found the bug there, and noticed our loading wasn't working and assumed the root cause was the same. It seems `get` returns something that does fulfil `instanceof Promise`, unlike `set`.
